addons/hr/models/hr_employee_base.py

```diff
@@ -70,10 +70,19 @@ class HrEmployeeBase(models.AbstractModel):
         ], string='Employee Type', default='employee', required=True,
         help="The employee type. Although the primary purpose may seem to categorize employees, this field has also an impact in the Contract History. Only Employee type is supposed to be under contract and will have a Contract History.")
 
+
+    def _get_valid_employee_for_user(self):
+        user = self.env.user
+        user_employee = user.employee_id or user.employee_id.sudo().search([
+            ('user_id', '=', self.env.user.id)
+        ], limit=1)
+        return user_employee
```

Comment:
@wtaferner what do think if we do like this? :)

```suggestion
        return self.env.user.sudo().employee_ids[0]
```

Reply:
Not sure as I think I have been there and had to go the other way 😄 and besides in case you have a proper employee in the current company it would not be correct and to be precise we would probably need to retrieve the employee for the default company set.

Do you agree with the approach itself? 

I have stumbled several times already over such awkward situations in the code where no employee is an incorrect assumptions. Do we already have a centralized method to retrieve some kind of fallback employee or would this be the start?

Reply:
Hey again, let's just state a problem and what we think should be a correct behavior and then we can discuss how to solve it.
To make sure we are on the same page :smile: 


Let's say we have a user and we have corresponding companies and employees

company                  |     employee

company _main       |     xxx
company _A             |     employee_A
company _B             |     employee_B
company _C             |     employee_C
company _D             |       xxx
company _E             |     employee_E


So, the our user (for some reason) does not have an employee in their main company. 
And this is when we get an issue, as `self.env.user.employee_id `is empty and this is what we
trying to solve, right?

How, let's say from the multi-company widget I have ticked in company_B and company_C
and I am using "My department" filter.

With sudo() and search combination you will override multi-company rule, so you will get
an employee regardless if their company is ticked or not, and as 'hr.employee' is ordered
by name, you will probably get Employee_A but because of multi-company rules on employee you
won't see employees that belong to company _A.
So, I guess, if `self.env.user.employee_id` is empty, we would like to get for the user one of the employees,
for witch the company is ticked from the multi-company widget.

My guess will be that in this case we would like to see either employee_B or employee_C.
And that will be done with self.env.user.employee_ids[0]. Even if we write sudo() in it,
like self.env.user.sudo()employee_ids we won't see all the employees anyways, to my knowledge.

So,you were right, my solution is not optimal as if `self.env.user.employee_id` is not empty,
we might not get it. 

How about 

```
user = self.env.user
return user.employee_id or user.employee_ids[0]
```

As for systematic fix, I don't know. There could be a business logic that we want employee of
the main company and that's it. Could be that for some logic, we don't want a fallback, and if I want, I could decide what I want my fallback to be. It's also not super obvious what should be fallback to all the scenarios, to me.

Reply:
I think you almost got the issue right, so basically I agree. I try to communicate it from my PoV and draw the solution or probably force-push that you can agree by merging it or challenging again.

Company 1 (Default for User A) ------ NOT TICKED --  No Employee 
Company 2 (Optional for User A)  --- NOT TICKED --  Employee B
Company 3 (Optional for User A) ---- NOT TICKED -- Employee C (sometimes employed in two companies, but rarely) 
Company 4 (Optional for User A) --- ACTIVE ---------- No Employee

- User A is active in Company 4 we will get no employee `self.env.user.employee_id` 
- First fallback I think should be the employee of the Default Company which is Company 1
- Second fallback should be the first employee found for the rest, depending on the order it would be the employee from company 2 or 3.

Due to multi company rules it seems that the environment needs to be tricked which I did already with limit 1.
Now I think we should remove that and try to filter for the default company of the user first and if no employee take the first one of the original result.

```python
user = self.env.user
employee = user.employee_id
if not employee:
    user_employees = user.employee_id.sudo().search([
        ('user_id', '=', user.id)
    ])
    employee = user_employees.filtered(lambda r: r.company_id == user.company_id) or user_employees[0]
return employee
```

And yes, I think this should be the fallback order in general in case it is allowed to fallback on an employee in another company to get the correct result for whatever we want to do.

Reply:
You have a point, indeed, even if the company is not ticked, we should be getting an employee associated with it. 

Code LGTM,

thanks for your contribution :raised_hands:

Reply:
Thank you for your time, review and a productive discussion to get the best result possible after that.